README
----PREREQUISITES----
*a) Install the MTi USB Serial Driver - follow the instructions below
git clone https://github.com/xsens/xsens_mt.git
cd ~/xsens_mt
make
sudo modprobe usbserial
sudo insmod ./xsens_mt.ko
*b) Install gps_common
sudo apt-get install ros-distro-gps-common
_____________________________________________________________________
----RUNNING XSENS MTi ROS NODE v1.0----
1. Copy the contents of the src folder into your catkin workspace 'src' folder. Change the permission to 'rw' using the chmod command for these folders
& folders. For details on creating a catkin workspace environment refer to "Creating a catkin ws" here
http://wiki.ros.org/ROS/Tutorials/InstallingandConfiguringROSEnvironment#Create_a_ROS_Workspace

2. in your catkin_ws folder, execute
catkin_make

3. source ~/catkin_ws/devel/setup.bash (Note: you have to do this for each terminal you open, if you are working in a single environment it is better to
source it automatically by editing your ~/.bashrc - add the line to end of the file)

4. Initiate 
roscore

5. Open a new terminal, type
rosrun xsens_driver mtdevice.py -q fs (fs can be 1,5,10,20,40,50,80,100,200 or 400Hz)
rosrun xsens_driver mtdevice.py -t fs (fs can be 1,5,10,20,40,50,80,100,200 or 400Hz)
e.g.
rosrun xsens_driver mtdevice.py -q 100 [this publishes the /xsens/sensorSample]
rosrun xsens_driver mtdevice.py -t 100 [this publishes the /imu/data]

This configures the MTi (MTi-10/20/30/100/200/300 or MTi-G-700) to output inertial data and magnetometer data at the set ODR. The maximum supported
inertial update rate is 400Hz and for the magnetometer it is 100Hz

6. To run the node
rosrun xsens_driver mtnode.py _device:=/dev/ttyUSB0 _baudrate:=115200
or
rosrun xsens_driver mtnode.py 

7. Open a new terminal (do not forget step 3)
rostopic echo /xsens/sensorSample  or rostopic echo /imu/data



CHANGELOG.md
# Changelog for package xsens_mti_ros_node

## 4.0.2 (2018-07-31)
* Updated gps package required in prerequisites section of README

## 4.0.1 (2018-07-20)
* Fixed typo in xsens.yaml file w.r.t MTi-3 filter profiles
* Changed example message in ReadMe

## 4.0.0 (2018-07-04)
* Merge pull request #33: Change custom_msgs to xsens_msgs
* Merge pull request #32: Configure MTi device on startup
* Added support to MTi-7 device

## 3.0.2 (2018-04-24)
* Added warning (rospy.logwarn) for timeouts

## 3.0.1 (2017-08-09)
* Support added to the node for operation with Kinetic environment

## 3.0.0 (2016-05-27)
* Added functionality to detect devices based on Product masks (incl. 1-series and FMT1000 devices)
* Increased the 'additionalTimeOutOffset' from 6ms to 10ms as default

## 2.0.2 (2015-11-23)
* Fix Readme

## 2.0.1 (2015-03-19)
* Fix documentation, switch to Markdown

## 2.0.0 (2015-03-19)
* Updated GNSS data identifiers
* Fixed a timeout issue
* Added new message files
* Add baudrate checks
* Add filter options
* Removed support for legacy devices

## 1.0.0 (2014-09-02)
* Improved on MK4 functionality to publish /xsens/sensorSample messages

Built on ethzasl_xsens_driver developed by previous contributors - Enrique Fernandez, Francis Colas, Paul Mathieu, Sam Pfeiffer, 
Benjamin Hitov, Francis Colas, Nikolaus Demmel, Stéphane Magnenat, fcolas.


src/xsens_driver/src/mtdef.py
"""Constant and messages definition for MT communication."""


class MID:
	"""Values for the message id (MID)"""
	## Error message, 1 data byte
	Error = 0x42
	ErrorCodes = {
		0x03: "Invalid period",
		0x04: "Invalid message",
		0x1E: "Timer overflow",
		0x20: "Invalid baudrate",
		0x21: "Invalid parameter"
	}

	# State MID
	## Wake up procedure
	WakeUp = 0x3E
	## Switch to config state
	GoToConfig = 0x30
	## Switch to measurement state
	GoToMeasurement = 0x10
	## Reset device
	Reset = 0x40

	# Informational messages
	## Request device id
	ReqDID = 0x00
	## DeviceID, 4 bytes: HH HL LH LL
	DeviceID = 0x01
	## Compatibility for XBus Master users
	InitMT = 0x02
	InitMTResults = 0x03
	## Request product code in plain text
	ReqProductCode = 0x1C
	## Product code (max 20 bytes data)
	ProductCode = 0x1D
	## Request firmware revision
	ReqFWRev = 0x12
	## Firmware revision, 3 bytes: major minor rev
	FirmwareRev = 0x13
	## Request data length according to current configuration
	ReqDataLength = 0x0A
	## Data Length, 2 bytes
	DataLength = 0x0B
	## Request GPS status (MTi-G only)
	ReqGPSStatus = 0xA6
	## GPS status (MTi-G only)
	GPSStatus = 0xA7

	# Device specific messages
	## Baudrate, 1 byte
	SetBaudrate = 0x18
	## Error mode, 2 bytes, 0000, 0001, 0002, 0003 (default 0001)
	SetErrorMode = 0xDA
	## Location ID, 2 bytes, arbitrary, default is 0
	SetLocationID = 0x84
	## Restore factory defaults
	RestoreFactoryDef = 0x0E
	## Transmit delay (RS485), 2 bytes, number of clock ticks (1/29.4912 MHz)
	SetTransmitDelay = 0xDC

	# Synchronization messages
	## Synchronization settings (MTi-10/100 series only), N*12 bytes
	SetSyncSettings = 0x2C
	## SyncIn setting (MTi only), (1+) 2 or 4 bytes depending on request
	SetSyncInSettings = 0xD6
	## SyncOut setting (MTi/MTi-G only), (1+) 2 or 4 bytes depending on request
	SetSyncOutSettings = 0xD8

	# Configuration messages
	## Request configuration
	ReqConfiguration = 0x0C
	## Configuration, 118 bytes
	Configuration = 0x0D
	## Output configuration (MTi-10/100 series only), N*4 bytes
	SetOutputConfiguration = 0xC0
	## Sampling period (MTi/MTi-G only), 2 bytes
	SetPeriod = 0x04
	## Skip factor (MTi/MTi-G only), 2 bytes
	SetOutputSkipFactor = 0xD4
	## Object alignment matrix, 9*4 bytes
	SetObjectAlignment = 0xE0
	## Output mode (MTi/MTi-G only), 2 bytes
	SetOutputMode = 0xD0
	## Output settings (MTi/MTi-G only), 4 bytes
	SetOutputSettings = 0xD2

	# Data messages
	## Request MTData message (for 65535 skip factor)
	ReqData = 0x34
	## Legacy data packet
	MTData = 0x32
	## Newer data packet (MTi-10/100 series only)
	MTData2 = 0x36

	# XKF Filter messages
	## Heading (MTi only), 4 bytes
	SetHeading = 0x82
	## Reset orientation, 2 bytes
	ResetOrientation = 0xA4
	## Request UTC time from sensor (MTI-G and MTi-10/100 series)
	ReqUTCTime = 0x60
	## UTC Time (MTI-G and MTi-10/100 series), 12 bytes
	UTCTime = 0x61
	## Request the available XKF scenarios on the device
	ReqAvailableScenarios = 0x62
	## Available Scenarios
	AvailableScenarios = 0x63
	## Current XKF scenario, 2 bytes
	SetCurrentScenario = 0x64
	## Magnitude of the gravity used for the sensor fusion mechanism, 4 bytes
	SetGravityMagnitude = 0x66
	## Lever arm of the GPSin sensor coordinates (MTi-G and MTi-700 only), 3*4 bytes
	SetLeverArmGPS = 0x68
	## Magnetic declination (MTi-G only), 4 bytes
	SetMagneticDeclination = 0x6A
	## Latitude, Longitude and Altitude for local declination and gravity
	# (MTi-10/100 series only), 24 bytes
	SetLatLonAlt = 0x6E
	## Processing flags (not on firmware 2.2 or lower for MTi/MTi-g), 1 byte
	SetProcessingFlags = 0x20
	## Initiate No Rotation procedure (not on MTi-G), 2 bytes
	SetNoRotation = 0x22
	
	## Some timeout related stuff
	additionalTimeOutOffset = 0.010 # 6ms


def getName(cls, value):
	'''Return the name of the first found member of class cls with given
	value.'''
	for k, v in cls.__dict__.iteritems():
		if v==value:
			return k
	return ''


def getMIDName(mid):
	'''Return the name of a message given the message id.'''
	name = getName(MID, mid)
	if name:
		return name
	if mid&1:
		name = getName(MID, mid-1)
		if name:
			return name+'Ack'
	return 'unknown MID'


class Baudrates(object):
	"""Baudrate information and conversion."""
	## Baudrate mapping between ID and value
	Baudrates = [
		(0x80, 921600),
		(0x0A, 921600),
		(0x00, 460800),
		(0x01, 230400),
		(0x02, 115200),
		(0x03,  76800),
		(0x04,  57600),
		(0x05,  38400),
		(0x06,  28800),
		(0x07,  19200),
		(0x08,  14400),
		(0x09,   9600),
		(0x0B,   4800),
		(0x80, 921600)]
	@classmethod
	def get_BRID(cls, baudrate):
		"""Get baudrate id for a given baudrate."""
		for brid, br in cls.Baudrates:
			if baudrate==br:
				return brid
		raise MTException("unsupported baudrate.")
	@classmethod
	def	get_BR(cls, baudrate_id):
		"""Get baudrate for a given baudrate id."""
		for brid, br in cls.Baudrates:
			if baudrate_id==brid:
				return br
		raise MTException("unknown baudrate id.")
	

class XDIGroup:
	"""Values for the XDI groups."""
	Temperature				= 0x0800
	Timestamp				= 0x1000
	OrientationData			= 0x2000
	Pressure				= 0x3000
	Acceleration			= 0x4000
	Position				= 0x5000
	AngularVelocity			= 0x8000
	GNSS					= 0x7000
	SensorComponentReadout	= 0xA000
	AnalogIn				= 0xB000
	Magnetic				= 0xC000
	Velocity				= 0xD000
	Status					= 0xE000

class XDIMessage:
	"""Values for the MKIV output data presets."""
	PacketCounter			= 0x00001020
	PaddedFs				= 0x0000FFFF
	SampleTimeFine			= 0x00001060
	DeltaV					= 0x00004010 
	DeltaVFs				= 0x00000190 # 400Hz
	FsModule				= 0x00000064 # 100Hz for 1-series and FMT1000 series
	Acceleration			= 0x00004020
	AccelerationFs 			= 0x00000190 # 400Hz
	DeltaQ					= 0x00008030
	DeltaQFs				= 0x00000190 # 400Hz
	RateOfTurn				= 0x00008020
	RateOfTurnFs			= 0x00000190 # 400Hz  
	MagneticField			= 0x0000C020
	MagneticFieldFs			= 0x00000064 # 100Hz
	Pressure				= 0x00003010
	PressureFs				= 0x00000032 # 50Hz
	StatusWord				= 0x0000E020
	GnssPvtData				= 0x00007010 # 4Hz TBD
	GnssSatInfo				= 0x00007020 # 4Hz TBD
	GnssFs					= 0x00000004
	PositionLatLon			= 0x00005040 # Latitude and longitude
	PositionHeight			= 0x00005020 # Ellipsoidal height
	Velocity 				= 0x0000D010 # Velocity in ENU
	Orientation				= 0x00002030 # Euler orientation ENU
	OrientationQuat			= 0x00002010 # Quaternion orientation ENU

class XDIProductMask:
	"""Product masks for the Xsens MTi series devices."""
	MTi1Series				= "8" # 0x08
	MTi7Device				= "7" # 0x07
	MTi10Series				= "6" # 0x06
	MTi100Series			= "7" # 0x07
	MTi700Device			= "7" # this is a subset of the 100-series 
	
	

class MTException(Exception):
	def __init__(self, message):
		self.message = message
	def __str__(self):
		return "MT error: " + self.message



src/xsens_driver/src/mtnode.py
#!/usr/bin/env python
import roslib; roslib.load_manifest('xsens_driver')
import rospy
import select
import mtdevice
import math
import pdb

from std_msgs.msg import Header, Float32, Float64
from sensor_msgs.msg import Imu
from geometry_msgs.msg import TwistStamped, Vector3Stamped, QuaternionStamped
from gps_common.msg import GPSFix, GPSStatus
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus, KeyValue
from xsens_msgs.msg import sensorSample, baroSample, gnssSample
from xsens_msgs.msg import positionEstimate, velocityEstimate, orientationEstimate

# transform Euler angles or matrix into quaternions
from math import pi, radians
from tf.transformations import quaternion_from_matrix, quaternion_from_euler, identity_matrix

import numpy

def get_param(name, default):
	try:
		v = rospy.get_param(name)
		rospy.loginfo("Found parameter: %s, value: %s"%(name, str(v)))
	except KeyError:
		v = default
		rospy.logwarn("Cannot find value for parameter: %s, assigning "
				"default: %s"%(name, str(v)))
	return v

class XSensDriver(object):

	ENU = numpy.identity(3)
	NED = numpy.array([[0, 1, 0], [ 1, 0, 0], [0, 0, -1]])
	NWU = numpy.array([[0, 1, 0], [-1, 0, 0], [0, 0,  1]])

	def __init__(self):

		device = get_param('~device', 'auto')
		baudrate = get_param('~baudrate', 0)
		if device=='auto':
			devs = mtdevice.find_devices()
			if devs:
				device, baudrate = devs[0]
				rospy.loginfo("Detected MT device on port %s @ %d bps"%(device,
						baudrate))
			else:
				rospy.logerr("Fatal: could not find proper MT device.")
				rospy.signal_shutdown("Could not find proper MT device.")
				return
		if not baudrate:
			baudrate = mtdevice.find_baudrate(device)
		if not baudrate:
			rospy.logerr("Fatal: could not find proper baudrate.")
			rospy.signal_shutdown("Could not find proper baudrate.")
			return

		rospy.loginfo("MT node interface: %s at %d bd."%(device, baudrate))
		self.mt = mtdevice.MTDevice(device, baudrate)

		configure_on_startup = get_param('~configure_on_startup', False)
		odr = get_param('~odr', 40)
		output_mode = get_param('~output_mode', 2)
		xkf_scenario = get_param('~xkf_scenario', 43)
		self.use_rostime = get_param('~use_rostime', False)

		if configure_on_startup:
			rospy.loginfo('Setting ODR (%d) and output mode (%d)' % (odr, output_mode))
			if odr not in [1, 2, 5, 10, 20, 40, 50, 80, 100, 200, 400]:
				raise Exception('Invalid ODR configuraton requested')
			if output_mode not in [1, 2, 3]:
				raise Exception('Invalid output mode requested')
			self.mt.configureMti(odr, output_mode)
			self.mt.ChangeBaudrate(baudrate)
			self.mt.SetCurrentScenario(xkf_scenario)
			self.mt.GoToMeasurement()
		else:
			rospy.loginfo('Using saved odr and output configuration')

		self.frame_id = get_param('~frame_id', 'mti/data')

		frame_local     = get_param('~frame_local'    , 'ENU')
		frame_local_imu = get_param('~frame_local_imu', 'ENU')

		if   frame_local == 'ENU':
			R = XSensDriver.ENU
		elif frame_local == 'NED':
			R = XSensDriver.NED
		elif frame_local == 'NWU':
			R = XSensDriver.NWU

		if   frame_local_imu == 'ENU':
			R_IMU = XSensDriver.ENU
		elif frame_local_imu == 'NED':
			R_IMU = XSensDriver.NED
		elif frame_local_imu == 'NWU':
			R_IMU = XSensDriver.NWU

		self.R = R.dot(R_IMU.transpose())

		self.diag_pub = rospy.Publisher('/diagnostics', DiagnosticArray, queue_size=10)
		self.diag_msg = DiagnosticArray()
		self.stest_stat = DiagnosticStatus(name='mtnode: Self Test', level=1,
				message='No status information')
		self.xkf_stat = DiagnosticStatus(name='mtnode: XKF Valid', level=1,
				message='No status information')
		self.gps_stat = DiagnosticStatus(name='mtnode: GPS Fix', level=1,
				message='No status information')
		self.diag_msg.status = [self.stest_stat, self.xkf_stat, self.gps_stat]

		self.imu_pub = rospy.Publisher('mti/sensor/imu', Imu, queue_size=10) #IMU message
		self.ss_pub = rospy.Publisher('mti/sensor/sample', sensorSample, queue_size=10) # sensorSample
		self.mag_pub = rospy.Publisher('mti/sensor/magnetic', Vector3Stamped, queue_size=10) # magnetic
		self.baro_pub = rospy.Publisher('mti/sensor/pressure', baroSample, queue_size=10) # baro
		self.gnssPvt_pub = rospy.Publisher('mti/sensor/gnssPvt', gnssSample, queue_size=10) # GNSS PVT
		#self.gnssSatinfo_pub = rospy.Publisher('mti/sensor/gnssStatus', GPSStatus, queue_size=10) # GNSS SATINFO
		self.ori_pub = rospy.Publisher('mti/filter/orientation', orientationEstimate, queue_size=10) # XKF/XEE orientation
		self.vel_pub = rospy.Publisher('mti/filter/velocity', velocityEstimate, queue_size=10) # XKF/XEE velocity
		self.pos_pub = rospy.Publisher('mti/filter/position', positionEstimate, queue_size=10) # XKF/XEE position
		
		self.temp_pub = rospy.Publisher('temperature', Float32, queue_size=10)	# decide type
		



	def spin(self):
		try:
			while not rospy.is_shutdown():
				self.spin_once()
		# Ctrl-C signal interferes with select with the ROS signal handler
		# should be OSError in python 3.?
		except select.error:
			pass

	def spin_once(self):
		
		def baroPressureToHeight(value):
			c1 = 44330.0
			c2 = 9.869232667160128300024673081668e-6
			c3 = 0.1901975534856
			intermediate = 1-math.pow(c2*value, c3)
			height = c1*intermediate
			return height
		
		# get data
		data = self.mt.read_measurement()
		# common header
		h = Header()
		h.stamp = rospy.Time.now()
		h.frame_id = self.frame_id

		# get data (None if not present)
		#temp = data.get('Temp')	# float
		orient_data = data.get('Orientation Data')
		velocity_data = data.get('Velocity')
		position_data = data.get('Latlon')
		altitude_data = data.get('Altitude')
		acc_data = data.get('Acceleration')
		gyr_data = data.get('Angular Velocity')
		mag_data = data.get('Magnetic')
		pressure_data = data.get('Pressure')
		time_data = data.get('Timestamp')
		gnss_data = data.get('Gnss PVT')
		status = data.get('Status')	# int

		# create messages and default values
		"Imu message supported with Modes 1 & 2"
		imu_msg = Imu()
		pub_imu = False
		"SensorSample message supported with Mode 2"
		ss_msg = sensorSample()
		pub_ss = False
		"Mag message supported with Modes 1 & 2"
		mag_msg = Vector3Stamped()
		pub_mag = False
		"Baro in meters"
		baro_msg = baroSample()
		pub_baro = False
		"GNSS message supported only with MTi-G-7xx devices"
		"Valid only for modes 1 and 2"
		gnssPvt_msg = gnssSample()
		pub_gnssPvt = False
		#gnssSatinfo_msg = GPSStatus()
		#pub_gnssSatinfo = False		
		# All filter related outputs
		"Supported in mode 3"
		ori_msg = orientationEstimate()
		pub_ori = False
		"Supported in mode 3 for MTi-G-7xx devices"
		vel_msg = velocityEstimate()
		pub_vel = False
		"Supported in mode 3 for MTi-G-7xx devices"
		pos_msg = positionEstimate()
		pub_pos = False
		
		secs = 0
		nsecs = 0
		
		if time_data:
			# first getting the sampleTimeFine
			time = time_data['SampleTimeFine']
			secs = 100e-6*time
			nsecs = 1e5*time - 1e9*math.floor(secs)							
		
		if acc_data:
			if 'Delta v.x' in acc_data: # found delta-v's
				pub_ss = True
				ss_msg.internal.imu.dv.x = acc_data['Delta v.x']
				ss_msg.internal.imu.dv.y = acc_data['Delta v.y']
				ss_msg.internal.imu.dv.z = acc_data['Delta v.z']											
			elif 'accX' in acc_data: # found acceleration
				pub_imu = True
				imu_msg.linear_acceleration.x = acc_data['accX']
				imu_msg.linear_acceleration.y = acc_data['accY']
				imu_msg.linear_acceleration.z = acc_data['accZ']						
			else:
				raise MTException("Unsupported message in XDI_AccelerationGroup.")	
					
		if gyr_data:
			if 'Delta q0' in gyr_data: # found delta-q's
				pub_ss = True
				ss_msg.internal.imu.dq.w = gyr_data['Delta q0']
				ss_msg.internal.imu.dq.x = gyr_data['Delta q1']
				ss_msg.internal.imu.dq.y = gyr_data['Delta q2']
				ss_msg.internal.imu.dq.z = gyr_data['Delta q3']
			elif 'gyrX' in gyr_data: # found rate of turn
				pub_imu = True
				imu_msg.angular_velocity.x = gyr_data['gyrX']
				imu_msg.angular_velocity.y = gyr_data['gyrY']
				imu_msg.angular_velocity.z = gyr_data['gyrZ']
			else:
				raise MTException("Unsupported message in XDI_AngularVelocityGroup.")
		
		if mag_data:
			# magfield
			ss_msg.internal.mag.x = mag_msg.vector.x = mag_data['magX']
			ss_msg.internal.mag.y = mag_msg.vector.y = mag_data['magY']
			ss_msg.internal.mag.z = mag_msg.vector.z = mag_data['magZ']
			pub_mag = True
			
		if pressure_data:
			pub_baro = True
			height = baroPressureToHeight(pressure_data['Pressure'])
			baro_msg.height = ss_msg.internal.baro.height = height
		
		if gnss_data:
			pub_gnssPvt = True
			gnssPvt_msg.itow = gnss_data['iTOW']
			gnssPvt_msg.fix = gnss_data['fix']			
			gnssPvt_msg.latitude = gnss_data['lat']
			gnssPvt_msg.longitude = gnss_data['lon']
			gnssPvt_msg.hEll = gnss_data['hEll']
			gnssPvt_msg.hMsl = gnss_data['hMsl']
			gnssPvt_msg.vel.x = gnss_data['velE']
			gnssPvt_msg.vel.y = gnss_data['velN']
			gnssPvt_msg.vel.z = -gnss_data['velD']
			gnssPvt_msg.hAcc = gnss_data['horzAcc']
			gnssPvt_msg.vAcc = gnss_data['vertAcc']
			gnssPvt_msg.sAcc = gnss_data['speedAcc']
			gnssPvt_msg.pDop = gnss_data['PDOP']
			gnssPvt_msg.hDop = gnss_data['HDOP']
			gnssPvt_msg.vDop = gnss_data['VDOP']
			gnssPvt_msg.numSat = gnss_data['nSat']
			gnssPvt_msg.heading = gnss_data['heading']
			gnssPvt_msg.headingAcc = gnss_data['headingAcc']

		if orient_data:
			if 'Q0' in orient_data:
				pub_imu = True
				imu_msg.orientation.w = orient_data['Q0']
				imu_msg.orientation.x = orient_data['Q1']
				imu_msg.orientation.y = orient_data['Q2']
				imu_msg.orientation.z = orient_data['Q3']
			elif 'Roll' in orient_data:
				pub_ori = True
				ori_msg.roll = orient_data['Roll']
				ori_msg.pitch = orient_data['Pitch']
				ori_msg.yaw = orient_data['Yaw']				
			else:
				raise MTException('Unsupported message in XDI_OrientationGroup')

		if velocity_data:
			pub_vel = True
			vel_msg.velE = velocity_data['velX']
			vel_msg.velN = velocity_data['velY']
			vel_msg.velU = velocity_data['velZ']
										
		if position_data:
			pub_pos = True
			pos_msg.latitude = position_data['lat']
			pos_msg.longitude = position_data['lon']
			
		if altitude_data:
			pub_pos = True	
			tempData = altitude_data['ellipsoid']
			pos_msg.hEll = tempData[0]
			
		#if status is not None:
		#	if status & 0b0001:
		#		self.stest_stat.level = DiagnosticStatus.OK
		#		self.stest_stat.message = "Ok"
		#	else:
		#		self.stest_stat.level = DiagnosticStatus.ERROR
		# 		self.stest_stat.message = "Failed"
		#	if status & 0b0010:
		#		self.xkf_stat.level = DiagnosticStatus.OK
		#		self.xkf_stat.message = "Valid"
		#	else:
		#		self.xkf_stat.level = DiagnosticStatus.WARN
		#		self.xkf_stat.message = "Invalid"
		#	if status & 0b0100:
		#		self.gps_stat.level = DiagnosticStatus.OK
		#		self.gps_stat.message = "Ok"
		#	else:
		#		self.gps_stat.level = DiagnosticStatus.WARN
		#		self.gps_stat.message = "No fix"
		#	self.diag_msg.header = h
		#	self.diag_pub.publish(self.diag_msg)

		
		# publish available information
		if pub_imu:
			imu_msg.header = h
			if not self.use_rostime:
				#all time assignments (overwriting ROS time)
				imu_msg.header.stamp.secs = secs
				imu_msg.header.stamp.nsecs = nsecs	
			self.imu_pub.publish(imu_msg)
		#if pub_gps:
		#	xgps_msg.header = gps_msg.header = h
		#	self.gps_pub.publish(gps_msg)
		#	self.xgps_pub.publish(xgps_msg)
		if pub_mag:
			mag_msg.header = h
			self.mag_pub.publish(mag_msg)
		#if pub_temp:
		#	self.temp_pub.publish(temp_msg)
		if pub_ss:
			ss_msg.header = h
			if not self.use_rostime:
				#all time assignments (overwriting ROS time)
				ss_msg.header.stamp.secs = secs
				ss_msg.header.stamp.nsecs = nsecs	
			self.ss_pub.publish(ss_msg)
		if pub_baro:
			baro_msg.header = h
			if not self.use_rostime:
				#all time assignments (overwriting ROS time)
				baro_msg.header.stamp.secs = secs
				baro_msg.header.stamp.nsecs = nsecs	
			self.baro_pub.publish(baro_msg)
		if pub_gnssPvt:
			gnssPvt_msg.header = h
			if not self.use_rostime:
				#all time assignments (overwriting ROS time)
				baro_msg.header.stamp.secs = secs
				baro_msg.header.stamp.nsecs = nsecs	
			self.gnssPvt_pub.publish(gnssPvt_msg)										
		if pub_ori:
			ori_msg.header = h
			if not self.use_rostime:
				#all time assignments (overwriting ROS time)
				ori_msg.header.stamp.secs = secs
				ori_msg.header.stamp.nsecs = nsecs	
			self.ori_pub.publish(ori_msg)
		if pub_vel:
			vel_msg.header = h
			if not self.use_rostime:
				#all time assignments (overwriting ROS time)
				vel_msg.header.stamp.secs = secs
				vel_msg.header.stamp.nsecs = nsecs	
			self.vel_pub.publish(vel_msg)
		if pub_pos:
			pos_msg.header = h
			if not self.use_rostime:
				#all time assignments (overwriting ROS time)
				pos_msg.header.stamp.secs = secs
				pos_msg.header.stamp.nsecs = nsecs	
			self.pos_pub.publish(pos_msg)		
			

def main():
	'''Create a ROS node and instantiate the class.'''
	rospy.init_node('xsens_driver', anonymous=True, log_level=rospy.INFO)
	driver = XSensDriver()
	driver.spin()


if __name__== '__main__':
	main()


